bob-access/src/credentials.rs

```diff
@@ -0,0 +1,159 @@
+use std::net::{IpAddr, SocketAddr};
+
+use axum::{
+    async_trait,
+    extract::{FromRequest, RequestParts},
+};
+use tonic::Request;
+
+use crate::{error::Error, extractor::ExtractorExt};
+
+#[derive(Debug, Default, Clone)]
+pub struct Credentials {
+    address: Option<SocketAddr>,
+    kind: Option<CredentialsKind>,
+}
+
+#[derive(Debug, Clone)]
+pub enum CredentialsKind {
+    Basic { username: String, password: String },
+    Token(String),
+}
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum CredentialsType {
+    Stub,
+    Basic,
+    Token,
+}
+
+impl CredentialsType {
+    pub fn is_basic(&self) -> bool {
+        *self == CredentialsType::Basic
+    }
+
+    pub fn is_stub(&self) -> bool {
+        *self == CredentialsType::Stub
+    }
+}
+
+fn _credentials_type(tp: Option<CredentialsType>) -> CredentialsType {
+    static mut CREDENTIALS_TYPE: CredentialsType = CredentialsType::Stub;
```

Comment:
This is not a very good solution. I suggest to try to use generic parametrization to pass the enum with data required for auth (None, UserPassword, Token). This can be achieved by declaring `CredentialsHolder<A>` struct, that directly wraps `Credentials` (the only difference is the generic parameter `A` which designate the authenticator). On `Authenticator` trait we should add the const enum field, that indicates the data that needs to be extracted from request for this particular authenticator. Additionally `FromRequest<B>` should be implemented not for `Credentials` but for `CredentialsHolder<A>` struct, and `CredentialsHolder<A>` struct should be used as a parameter in all REST API methods. In this way extractor can get information about fields from `Authenticator` type and pass them to `ExtractorExt::extract`